[#148] Autocomple for 'Raw Fact' strings

elbenfreund opened the pull request:
This PR changes the way raw fact strings are handled by our GUI.
Instead of using a generic ``Gtk.Entry`` we introduce a new ``RawFactEntry`` class to handle the specific requirements of 'raw fact' strings. 
Autocompletion is provided by a customized ``EntryCompletion`` class called ``RawFactCompletion``.
Both custom classes are to be used in conjunction.

The basic approach to the problem at hand is as follows.

1. ``RawFactEntry._on_changed`` decomposes the current raw fact string and tries to map its substrings to activity, category, tags, etc... segments. This is done with the help of regex matching.
2. If the entry text could be mapped at all, the entry uses the current cursor position to make a guess what segment the user is currently trying to specify.
3. If the currently edited segment is one that we want to provided autocompletion for, the appropiate model is set on the completion instance and the completion machinery is launched.
4. If the user selects a completion suggestion, the original segment text gets replaced by the completion text and the cursor placed at the segment end.

There are quite some subtleties involved as the result of our non standard matching/autocomplete requirements applying only to substrings of the entry (for more details see the commit msg and code annotations) but this is the gist of it.

One central cornerstone of this solution is our new 'raw fact' regex. This entire approach depends on our ability to come up with a regex that is able to capture the semantics of our raw fact string sufficiently well.
Please not that this does not include any validation at all.
*If* this works out, we may even have a good candidate to refactor ``hamster_lib.Fact.create_from_raw_fact`` to use this regex based solution instead of its current rather hacky one.

Closes: #148

elbenfreund committed:
Minor docstring and import fixes

elbenfreund committed:
Pass controller instead of app to TrackingScreen

TrackingScreens signature has been adjusted to reflect the fact that it
only needs access to the controller instance and not the whole app.

elbenfreund committed:
Add RawFactEntry and Completion

This commit introduces a new helper ``helpers.decompose_raw_fact_string`` which is
responsible for disecting a given 'raw fact' string and extracting as
much information as possible. Unlike the ``hamster-lib`` function of
similar spirit this one takes a straight forward regex based approach to the
problem. As such it allows to return either the raw ``match`` instance
or the corresponding groupdict.

We introduce a new custom entry widget: ``RawFactEntry``. Its main
purpose is to encapsulate all raw fact related user  interaction such as
autocompletion etc. Finally we add ``RawFactCompletion`` as a custom
``Gtk.EntryCompletion`` class that is responsible for actually providing
autocompletion.

Closes: #148

jtojnar commented:
At a glance, it looks reasonable. I am mostly concerned about the `decompose_raw_fact_string` helper.
1. Descriptions cannot contain at-signs or hashes – `foo@bar, @baz` or `foo@bar, #baz` do not match anything.
2. Tags can contain at-signs, though – `foo #@bar` does match.
3. Relative times are not matched.
4. Category without activity can be matched.
5. The function returns the matches including separators and spaces.

elbenfreund commented:
You are totally right to be concerned about `decompose_raw_fact_string`. This clearly carries a lot of weight and is hard to get right.
What I did here is just a first and relatively restrictive approximation. By composing the regex in such a modular way we should be able to experiment quite easily to hopefully find something that works nicely.
The challenge is to create a regex that matches valid strings, but also is powerfull and flexible enough to have a good change of matching incomplete strings sufficiently well.
The legacy parsing has its own issues and by now I would be willing to break compatibility if that would be really needed for a clear, concise and transparent raw fact string semantic (for example, one could consider using `;` as a description seperator instead.

Now on to you first feedback.
1. Yes, and I agree that it would be good to allow as many characters in the description (including `@` and `#` as possible. I will add a commit that will give this a try. There is always the risk of side effects but this is something only experience will show.
2. That is deliberate: as I said, I try to limit the char-set as little as possible wherever possible.
3. Yes, thats something still to be done. Thanks for the reminder.
4. That is intensional. As this is not a `create from raw fact` but rather a 'get me all the information contained in the string'  function this is exactly the way we want it to behave.
5. Yes. I wanted to make sure that the function makes as little preassumptions about the way it is used as possible. Right now `RawFactEntry` provides its own prefix removal and alllows easy access to the cleaned up segment text via `RawFactEntry.get_segment_text`. If this a recurring issue, prefix removal could be outsourced to a seperate helper method that provides cleaned up strings.

elbenfreund committed:
General cleanup

Also: updated sphinx version

jtojnar commented on hamster_gtk/helpers.py:
```diff
@@ -137,3 +138,62 @@ def get_config_instance(fallback_config_instance, app_name, file_name):
+
+    Args:
+        text (text_type): String to be analysed.
+        raw (bool): If True, return the raw match instance, if ``False`` return
```
`False` is backticked here, while `True` is not.

jtojnar commented on README.rst:
```diff
@@ -33,7 +33,7 @@ To Run the Testsuite
 * Install dependencies (on debian if using virtualenvwrapper):
-  ``apt-get install virtualenvwrapper python-gi gir1.2-gtk-3.0 libglib2.0-dev
+  install virtualenvwrapper python-gi gir1.2-gtk-3.0 libglib2.0-dev
   libgtk-3-dev``.
```
Forgotten closing backticks.

jtojnar commented on hamster_gtk/helpers.py:
```diff
@@ -137,3 +138,62 @@ def get_config_instance(fallback_config_instance, app_name, file_name):
+        This is not at all about providing valid facts or even raw facts. This function
+        is only trying to extract whatever information can be matched to its various
+        groups (aka 'segments').
+        This not withstanding can however be the bases for future implementations
```
Notwithstanding is a single word, however is probably redundant, and bases should be basis.

jtojnar commented on hamster_gtk/misc/widgets/raw_fact_entry.py:
```diff
@@ -0,0 +1,229 @@
+
+
+class RawFactEntry(Gtk.Entry):
+    """A custom entry widgets that provides ``faw fact`` specific autocompletion behaviour."""
```
widgets → widget, faw → raw

jtojnar commented on hamster_gtk/helpers.py:
```diff
@@ -137,3 +138,62 @@ def get_config_instance(fallback_config_instance, app_name, file_name):
+    # This is the central place where we define which characters are viable for
+    # our various segments.
+    # Please note that this is also where we define each segments 'separator'.
+    activity_regex = r'[^@:]+'
```
I am still not fully convinced this is the way to go. [Partial matches](https://pypi.python.org/pypi/regex) could be also considered.

jtojnar replied:
Additionally, I think `a#b@c` should be parsed as `{'activity': 'a', 'tags': '#b@c'}`, that is `#` should be banned from `activity`.

elbenfreund replied:
Is there a particular argument to be had for "banning ``#`` within the activity name"?
I personally do not care either way, but I think a general rule "allow as much as possible, unless for technical reasons" makes sense?

jtojnar replied:
Otherwise, `a #b` would be ambiguous, I think.

elbenfreund replied:
that should be ``Activity.name='a', Activity.tags={b}``, or not?

jtojnar replied:
I think this regex parses it as `Activity(name='a #b')`.

jtojnar commented on hamster_gtk/misc/widgets/raw_fact_entry.py:
```diff
@@ -0,0 +1,229 @@
+
+        match = self.match
+        segment = self.current_segment
+        if add_prefix:
```
This will always be `True` as the argument is shadowed by the helper function.

jtojnar commented on hamster_gtk/misc/widgets/raw_fact_entry.py:
```diff
@@ -0,0 +1,229 @@
+        self.current_segment = None
+        self.connect('changed', self._on_changed)
+
+    def replace_segment_string(self, segment_string, add_prefix=True):
```
What is the purpose of the `add_prefix` when it is never used?

jtojnar replied:
Also, should not this method by private?

elbenfreund replied:
I think it is a rather generic method and thought to make it part of the public API. Which is also why I included the ``add_prefix`` in the signature - to increase user ease. But you are right, I can not really come up with a usecase where one would want to deal with prefixing itself (e.g. one would use ``add_prefix=False``). Damn you featuritis. Going to remove it.

jtojnar commented on hamster_gtk/misc/widgets/raw_fact_entry.py:
```diff
@@ -0,0 +1,229 @@
+        self.set_text(new_string)
+        self.set_position(match.start(segment) + len(segment_string))
+
+    def get_segment_text(self, remove_prefix=True):
```
What is the purpose of the `remove_prefix` when it is never used?

jtojnar replied:
Also, should not this method by private?

elbenfreund replied:
Same as ``replace_segment_string``

jtojnar commented on hamster_gtk/misc/widgets/raw_fact_entry.py:
```diff
@@ -0,0 +1,229 @@
+            return result
+
+        result = self.match.group(self.current_segment)
+        if remove_prefix:
```
The argument is shadowed by the function.

jtojnar commented on hamster_gtk/misc/widgets/raw_fact_entry.py:
```diff
@@ -0,0 +1,229 @@
+        category of `` #`` for tags.
+        """
+        def remove_prefix(segment, string):
+            result = string
```
In `add_prefix` this is in the else block, it should be unified.

elbenfreund replied:
Yes. I can never make up my mind which to find better.

jtojnar commented on hamster_gtk/misc/widgets/raw_fact_entry.py:
```diff
@@ -0,0 +1,229 @@
+
+
+    # Callbacks
+    def _on_facts_changed(self, evtl):
```
evtl?

jtojnar commented on hamster_gtk/misc/widgets/raw_fact_entry.py:
```diff
@@ -0,0 +1,229 @@
+        Callback triggered whenever entry text is changed.
+
+        Its main task is to keep track of which segment of the raw fact string the
+        user is currently editing. For this the whole string is inspecred and matched
```
inspecred

jtojnar commented on hamster_gtk/misc/widgets/raw_fact_entry.py:
```diff
@@ -0,0 +1,229 @@
+
+        Its main task is to keep track of which segment of the raw fact string the
+        user is currently editing. For this the whole string is inspecred and matched
+        against our regex. By coparing the current cursor position with individual
```
coparing

jtojnar commented on hamster_gtk/misc/widgets/raw_fact_entry.py:
```diff
@@ -0,0 +1,229 @@
+    """
+
+    def __init__(self, controller, *args, **kwargs):
+        """instantiate class."""
```
Should begin with capital letter.

jtojnar commented on hamster_gtk/misc/widgets/raw_fact_entry.py:
```diff
@@ -0,0 +1,229 @@
+        # [FIXME]
+        # This needs to be improved. Right now activity and categories can be
+        # duplicates.
+        activities, categories = [], []
```
Why not just use set again?

elbenfreund replied:
The original reason was to keep the "order information" so one could use it to prioritize matches. But if we ever implement a more clever matching solution (may in line with legacy hamster) we can just remove the ``set``. So you are right, as it is now its neither here nor there...

jtojnar commented on hamster_gtk/misc/widgets/raw_fact_entry.py:
```diff
@@ -0,0 +1,229 @@
+        """
+        Return all activities that should be considered for autocompletion.
+
+        This is the place were we define which reference frame should be used
```
were → where

jtojnar commented:
Additionally, the separate autocomplete for activities and categories makes no sense to me. After all, each activity has a fixed category. I like how the code is structured, though; it will come in handy for tag autocompletion.

elbenfreund committed:
Improve docstrings and spelling

elbenfreund commented:
With regards to splitting autocomplete for activity and category:

Please not that it *is true* that each ``Actitivity`` has at max. one category assigned, but ``Activity.name`` ``is not`` a unique (natural) key! Only ``Activity.name + Activity.category`` as a composite key are unique.

Now, this may be something one could do different, but for now this is how ``hamster-lib`` (following legacy hamster for compatibility) handles things.

Now, that does not by itself mean that autocompletion could not behave differently to its current suggestion. For starters, when looking up categories, it may very well make sense to just look up categories that "match" the already known activity (instead of all fetched ones). Or one threats the "activity string" as the whole of "name + category" and looks it up as a whole.
At this point the given solution seemed to strike a balance between flexibility and implementation effort.
I think it would be good not to focus too much on user preferences but get the general machinery in place. Details like this can then be easily adapted to RL user expectations.

jtojnar commented:
Matching against `{name}@{category}` seems to be what hamster 2 is doing. It also feels more natural.

elbenfreund committed:
Refactor RawFactEntry and Completion

Remove ability to add/remove prefixes when getting segment text.
Minor clanup

elbenfreund committed:
Dissalow '#' within activity name

elbenfreund committed:
Disallow ',' in activity name

elbenfreund committed:
Refine docstrings and fix typos

elbenfreund committed:
Add autocompletion against full 'name+category'

This commit introduces a way to have the autocompletion match against
the full ``Activity.name@Activity.category.name`` string instead of
handeling both individually. The behaviour is controlled by
``RawFactEntry._split_activity_autocomplete``.
As part of this a new *pseudo-segment* called ``activity+category`` is
created that represents this combined version. As this new segment is
not actually part of our match object we have to introduce some
dedicated code switches. In order to limit code duplication, figuring
out a segments start and end position has been moved into a dedicated
private helper function ``_get_segment_boundaries`` as the old
``match.span`` based solution would no longer suffice.

elbenfreund committed:
Remove extra blank line and fix line length

jtojnar commented on hamster_gtk/misc/widgets/raw_fact_entry.py:
```diff
@@ -29,14 +29,50 @@
+    # We can not simply use ``match.span`` as ``activity+category`` is not
+    # represented as such in the match instance.
+    if segment == 'activity+category':
+        start = match.start('activity')
```
This should probably allow matching start of category when activity is not specified. At least this is how it behaves in hamster-master.

jtojnar commented on hamster_gtk/misc/widgets/raw_fact_entry.py:
```diff
@@ -29,14 +29,50 @@
+        match: A match object instance.
+
+    Returns:
+        tuple: Tuple (int, int) representing the given segments start and end position.
```
segment's

jtojnar commented on hamster_gtk/misc/widgets/raw_fact_entry.py:
```diff
@@ -29,14 +29,50 @@
-        """Instantiate class."""
+    def __init__(self, controller, split_activity_autocomplete=False, *args, **kwargs):
+        """
+        Instantiate class.
```
Should be initialize.

jtojnar commented on hamster_gtk/misc/widgets/raw_fact_entry.py:
```diff
@@ -95,8 +132,17 @@ def remove_prefix(segment, string):
+            return
+
+        if self.current_segment == 'activity+category':
+            result = self.match.group('activity')
```
To allow matching just on categories, something like `result = (self.match.group('activity') or '') + (self.match.group('category') or '')` should be used.

elbenfreund committed:
Fix typo

elbenfreund committed:
'activits+catagories' can match against just categories

If the decomposed raw fact does not include an 'activity' but a
'category' segment (e.g. "@foo") 'activity+category' will now be able to
deal with this and as a consequence will autocomplete against all
activities with this category. This allows for fast activity lookup if
the category is known but not the activity name.

jtojnar commented on hamster_gtk/misc/widgets/raw_fact_entry.py:
```diff
@@ -39,12 +39,14 @@ def _get_segment_boundaries(segment, match):
     # represented as such in the match instance.
     if segment == 'activity+category':
         start = match.start('activity')
+        if start == -1:
```
This should probably be the same as the second condition.

jtojnar commented on hamster_gtk/misc/widgets/raw_fact_entry.py:
```diff
@@ -136,13 +138,20 @@ def remove_prefix(segment, string):
+            result = self.match.group(self.current_segment)
+            if result:
+                result = remove_prefix(self.current_segment, result)
+            result
```
Forgotten line

elbenfreund commented:
Ready for you :)

elbenfreund committed:
Merge branch 'develop' into feature/148-Autocomple_for_raw_facts
